### PR TITLE
More arena cleanups

### PR DESCRIPTION
**Pull request by nnethercote**
A sequel to #90990.

r? @oli-obk

### CODE REVIEW

**Comment by nnethercote** [comment]
Best reviewed one commit at a time.

File: compiler/rustc_arena/src/lib.rs [hunk 1]
```diff
@@ -521,9 +526,18 @@ impl DroplessArena {
...
+// There are three cases of interest.
+// - Types that are `Copy`: these need not be specified in the arguments. They
+//   will use the `DroplessArena`.
+// - Types that are `!Copy` and `!Drop`: these must be specified in the
+//   arguments. An empty `TypedArena` will be created for each one, but the
+//   `DroplessArena` will always be used and the `TypedArena` will stay empty.
+//   This is odd but harmless, because an empty arena allocates no memory.
+// - Types that are `!Copy` and `Drop`: these must be specified in the
+//   arguments. The `TypedArena` will be used for them.
+//
```
Comment (bjorn3) [inline compiler/rustc_arena/src/lib.rs:540]: Should this become a doc comment?

Comment (nnethercote) [reply compiler/rustc_arena/src/lib.rs]: Good idea. I have updated it.

**Pull request merged**